test/test_specs.py

```diff
@@ -2348,6 +2348,112 @@ def test_to_numpy(self):
+    assert spec[[0]].shape == torch.Size([1, 4])
+    assert spec[[[[0]]]].shape == torch.Size([1, 1, 1, 4])
+    assert spec[[0, 1]].shape == torch.Size([2, 4])
+    assert spec[[[0, 1]]].shape == torch.Size([1, 2, 4])
+    assert spec_3d[[0, 1], [0, 1]].shape == torch.Size([2, 4])
+    assert spec[[[0, 1], [0, 1]]].shape == torch.Size([2, 2, 4])
+    # Tuples
+    assert spec_3d[1, 2].shape == torch.Size([4])
+    assert spec_3d[(1, 2)].shape == torch.Size([4])
+    assert spec_3d[((1, 2))].shape == torch.Size([4])
+    # Ranges
+    assert spec[range(2)].shape == torch.Size([2, 4])
+    # Slices
+    assert spec[:].shape == torch.Size([3, 4])
+    assert spec[10:].shape == torch.Size([0, 4])
+    assert spec[:1].shape == torch.Size([1, 4])
+    assert spec[1:2].shape == torch.Size([1, 4])
+    assert spec[10:1:-1].shape == torch.Size([1, 4])
+    assert spec[-5:-1].shape == torch.Size([2, 4])
+    assert spec_3d[[1, 2], 3:].shape == torch.Size([2, 0, 4])
+    # None (adds a singleton dimension where needed)
+    assert spec[None].shape == torch.Size([1, 3, 4])
+    assert spec[None, :2].shape == torch.Size([1, 2, 4])
+    expected_shape = [1, 0] if isinstance(spec, OneHotDiscreteTensorSpec) else [1]
+    assert empty_spec[None].shape == torch.Size(expected_shape)
+    # Ellipsis
+    expected_shape = [0] if isinstance(spec, OneHotDiscreteTensorSpec) else []
+    assert empty_spec[...].shape == torch.Size(expected_shape)
+    expected_shape = [2, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 2]
+    assert spec[..., :2].shape == torch.Size(expected_shape)
+    expected_shape = (
+        [2, 1, 1, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 2, 1, 1]
+    )
+    assert spec[..., :2, None, None].shape == torch.Size(expected_shape)
+    expected_shape = [3, 6] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 4]
+    assert spec_4d[1, ..., 2].shape == torch.Size(expected_shape)
+    assert spec[1, ...].shape == torch.Size([4])
+    expected_shape = [1, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [4, 1]
+    assert spec[1, ..., None].shape == torch.Size(expected_shape)
+    expected_shape = [2, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [5, 2]
+    assert spec_3d[..., [0, 1], [0]].shape == torch.Size(expected_shape)
+    expected_shape = (
+        [1, 3, 1, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [1, 3, 4, 1]
+    )
+    assert spec_3d[None, 1, ..., None].shape == torch.Size(expected_shape)
+    # Numpy arrays
+    assert spec[np.array([[1, 2]])].shape == torch.Size([1, 2, 4])
+    # Tensors
+    assert spec[torch.randint(3, (3, 2))].shape == torch.Size([3, 2, 4])
+    # Tuples
+    # Note: nested tuples are supported by specs but transformed into lists, similarity to numpy
+    assert spec_3d[(0, 1), (0, 1)].shape == torch.Size([2, 4])
+    assert spec_3d[:2, (0, 1)].shape == torch.Size([2, 2, 4])
+    assert spec_3d[:2, [0, 1]].shape == torch.Size([2, 2, 4])
```

Comment:
to test with a tensor:
```suggestion
    assert spec_3d[:2, (0, 1)].shape == torch.Size([2, 2, 4])
    assert spec_3d[:2, [0, 1]].shape == torch.Size([2, 2, 4])
    assert spec_3d[:2, torch.tensor([0, 1])].shape == torch.Size([2, 2, 4])
```

Reply:
Good call. Support was added for indexing of range, ndarrays and tensors within tuples.